x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -137,15 +138,15 @@ func (s *websocketStream) FollowStream(ctx context.Context) error {
 					s.log.Errorw("failed to read websocket data", "error", err)
 					return err
 				}
-			}
-			s.metrics.receivedBytesTotal.Add(uint64(len(message)))
-			state["response"] = message
-			s.log.Debugw("received websocket message", logp.Namespace("websocket"), "msg", string(message))
-			err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
-			if err != nil {
-				s.metrics.errorsTotal.Inc()
-				s.log.Errorw("failed to process and publish data", "error", err)
-				return err
+			} else {
+				state["response"] = message
+				s.log.Debugw("received websocket message", logp.Namespace("websocket"), "msg", string(message))
+				err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
+				if err != nil {
+					s.metrics.errorsTotal.Inc()
+					s.log.Errorw("failed to process and publish data", "error", err)
+					return err
+				}
```

Comment:
```suggestion
				continue
			}
			state["response"] = message
			s.log.Debugw("received websocket message", logp.Namespace("websocket"), "msg", string(message))
			err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
			if err != nil {
				s.metrics.errorsTotal.Inc()
				s.log.Errorw("failed to process and publish data", "error", err)
				return err
			}
```
but why does the error case not return if `!isRetryableError(err)`?

Reply:
@efd6, It does return, the issue is when it is a **RetryableError**. In this scenario the message is still passed to CEL during the retry attempts. So if the 3rd attempt is successful the 1st 2 invalid messages are still processed by CEL leading to downstream errors in integration pipelines because the event itself might be malformed or not present. Also this is just unnecessary processing that can be avoided.

Reply:
Also I felt that an else block was a clearer from a readability perspective since we already have a bunch of returns within the "if".

Reply:
> It does return, the issue is when it is a RetryableError

Sorry, you are correct.

Suggest then
```
			if err != nil {
				s.metrics.errorsTotal.Inc()
				if !isRetryableError(err) {
					s.log.Errorw("failed to read websocket data", "error", err)
					return err
				}
				s.log.Debugw("websocket connection encountered an error, attempting to reconnect...", "error", err)
				// close the old connection and reconnect
				if err := c.Close(); err != nil {
					s.metrics.errorsTotal.Inc()
					s.log.Errorw("encountered an error while closing the websocket connection", "error", err)
				}
				// since c is already a pointer, we can reassign it to the new connection and the defer func will still handle it
				c, resp, err = connectWebSocket(ctx, s.cfg, url, s.log)
				handleConnectionResponse(resp, s.metrics, s.log)
				if err != nil {
					s.metrics.errorsTotal.Inc()
					s.log.Errorw("failed to reconnect websocket connection", "error", err)
					return err
				}
				continue
			}
			state["response"] = message
			s.log.Debugw("received websocket message", logp.Namespace("websocket"), "msg", string(message))
			err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
			if err != nil {
				s.metrics.errorsTotal.Inc()
				s.log.Errorw("failed to process and publish data", "error", err)
				return err
			}
```
Note also that the debugw call with the message allocates non-conditionally due to the `string` conversion. This can be avoided by providing a lazy `fmt.Stringer`.

Reply:
I'll keep the fmt.Stringer change for a later cleanup PR then.

Reply:
The type is
```
type bytesStringer []byte

func (b bytesStringer) String() string { return string(b) }
```